### PR TITLE
Add support for JSON Schema structured output

### PR DESCRIPTION
**Pull request by daguava**
Ollama has [introduced support for structuring responses](https://github.com/ollama/ollama/blob/main/docs/api.md#request-3) as defined by a JSON Schema. Ollama, when provided a valid JSON Schema, will convert it into a grammar that ensures it will generate a response matching your specification.

This replaces the existing `format: Boolean` parameter Ollama had previously.

This PR introduces a feature to leverage this structured output. When generation / chat requests are made a class may be specified to define a structure in the OllamaGenerateRequest or OllamaChatRequest via `withResponseClass(MyClass.java)`. Now, when an OllamaResult is received and you had specified a class, calling `getStructuredResponse<MyClass>()` returns an instance of that class as expected.

The fun part of this happens behind the scenes. Before generation, the provided class has a JSON Schema representation generated from the class structure. This is the format that the Ollama server will use to structure the output to be returned, and ensures we can unmarshal the generated content back to this provided class in Java. The JSON Schema output is also cached for re-use.

To avoid changing existing contracts this was added with a new getter `getStructuredResponse`. 

In summary:
1) allow specifying a class to structure ollama output with in the generate/chat request
2) use it to generate a json schema to supply to ollama so that it follows the format
3) convert it back to an instance of the given class when calling getStructuredResponse()

If I've missed anything please let me know, I'm open to feedback and changes.

### CODE REVIEW

**Comment by daguava** [comment]
Update 2: This is working and I am actively using it in my own repository (via Kotlin)

```
internal inline fun <reified T> getStructuredResponse(
    system: TwitchChatMessage?,
    message: TwitchChatMessage,
): T? {
    val fullResponse =
        ollamaAPI.chat(
            OllamaChatRequestBuilder
                .getInstance(MODEL)
                .withMessages(
                    system?.let {
                        mutableListOf(
                            OllamaChatMessage(OllamaChatMessageRole.SYSTEM, it.content),
                        )
                    } ?: mutableListOf(),
                )
                .withMessage(
                    OllamaChatMessageRole.getRole(message.role),
                    message.content,
                )
                .withOptions(OptionsBuilder().setNumCtx(FULL_CONTEXT_LENGTH).build())
                .withResponseClass(T::class.java)
                .build(),
        )

    val textResponse = fullResponse.response
    val structuredResponse = fullResponse.getStructuredResponse<T>()
 ```

Which is then used later on...

            
```
currentQuestion = getStructuredResponse<Trivia?>(modifiedSystemPrompt, question)
```

With trivia being the kotlin equivalent of a java record

```
data class Trivia(
    val question: String = "",
    val answer: List<String> = emptyList(),
    val hint: String = "",
)
```

I'm really liking the automatic JSON schema generation approach a lot.

**Comment by daguava** [comment]
Broke it during the cleanup, one moment :)

**Comment by daguava** [comment]
I'm a little more used to Kotlin where you can use reified types for runtime type information so my java serialization knowledge isn't my strong suit - please let me know how this looks, I'm willing to make any desired changes :)

**Comment by amithkoujalgi** [comment]
@daguava there are conflicts in the code. Mind resolving them and updating the PR? Thanks!

**Comment by daguava** [comment]
> @daguava there are conflicts in the code. Mind resolving them and updating the PR? Thanks!

Yes, I'll update this later today to account for the other tools PR that merged.

**Comment by daguava** [comment]
Closing for now, I'll submit a new PR when I've had the time to merge this with the newest changes.
